curp/src/server/raw_curp/mod.rs

```diff
@@ -0,0 +1,806 @@
+    /// Candidate's term
+    pub(super) term: u64,
+    /// Candidate's Id
+    pub(super) candidate_id: ServerId,
+    /// Candidate's last log index
+    pub(super) last_log_index: usize,
+    /// Candidate's last log term
+    pub(super) last_log_term: u64,
+}
+
+/// Invoked by leader to replicate log entries; also used as heartbeat
+pub(super) struct AppendEntries<C> {
+    /// Leader's term
+    pub(super) term: u64,
+    /// Leader's id
+    pub(super) leader_id: ServerId,
+    /// Index of log entry immediately preceding new ones
+    pub(super) prev_log_index: usize,
+    /// Term of log entry immediately preceding new ones
+    pub(super) prev_log_term: u64,
+    /// Leader's commit index
+    pub(super) leader_commit: usize,
+    /// New entries to be appended to the follower
+    pub(super) entries: Vec<LogEntry<C>>,
+}
+
+/// Curp Role
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum Role {
+    /// Follower
+    Follower,
+    /// Candidate
+    Candidate,
+    /// Leader
+    Leader,
+}
+
+/// Relevant context for Curp
+struct Context<C: Command> {
+    /// Id of the server
+    id: ServerId,
+    /// Other server ids
+    others: HashSet<ServerId>,
+    /// Timeout
+    timeout: Arc<ServerTimeout>,
+    /// Cmd board for tracking the cmd sync results
+    cb: CmdBoardRef<C>,
+    /// Speculative pool
+    sp: SpecPoolRef<C>,
+    /// Tx to send leader changes
+    leader_tx: broadcast::Sender<Option<ServerId>>,
+    /// Election tick
+    election_tick: AtomicU8,
+    /// Heartbeat opt out flag
+    hb_opt: AtomicBool,
+    /// Tx to send cmds to execute and do after sync
+    cmd_tx: Box<dyn CmdExeSenderInterface<C>>,
+}
+
+impl<C: Command> Debug for Context<C> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Context")
+            .field("id", &self.id)
+            .field("others", &self.others)
+            .field("timeout", &self.timeout)
+            .field("cb", &self.cb)
+            .field("sp", &self.sp)
+            .field("leader_tx", &self.leader_tx)
+            .field("election_tick", &self.election_tick)
+            .field("hb_opt", &self.hb_opt)
+            .finish()
+    }
+}
+
+// Tick
+impl<C: 'static + Command> RawCurp<C> {
+    /// Tick
+    pub(super) fn tick(&self) -> TickAction<C> {
+        let (role, timeout) = self.st.map_read(|st_r| {
+            (
+                st_r.role,
+                if st_r.role == Role::Follower {
+                    st_r.follower_timeout_ticks
+                } else {
+                    st_r.candidate_timeout_ticks
+                },
+            )
+        });
+        if role == Role::Leader {
+            self.tick_heartbeat()
+        } else {
+            self.tick_election(timeout)
+        }
+    }
+
+    /// Tick election, will generate votes if timeout
+    fn tick_election(&self, timeout: u8) -> TickAction<C> {
+        let tick = self.ctx.election_tick.fetch_add(1, Ordering::SeqCst);
```

Comment:
Why should we use `Ordering::SeqCst` here? It seems that `Ordering::AcqReql` is enough.

Reply:
Fixed.